dbms/src/Storages/DeltaMerge/Segment.cpp

```diff
@@ -2224,5 +2255,285 @@ bool Segment::placeDelete(const DMContext & dm_context,
+{
+    Stopwatch sw;
+    const auto & dmfiles = segment_snap->stable->getDMFiles();
+    RUNTIME_CHECK(!dmfiles.empty(), dmfiles.size());
+    std::vector<std::vector<PackInfo>> all_packs{dmfiles.size()};
+    std::vector<IdSetPtr> some_packs{dmfiles.size()};
+    bool all_match = true;
+    size_t some_packs_count = 0;
+    for (size_t i = 0; i < dmfiles.size(); i++)
+    {
+        std::tie(all_packs[i], some_packs[i]) = parseDMFilePackInfo(dmfiles[i],
+                                                                    dm_context,
+                                                                    read_ranges,
+                                                                    filter,
+                                                                    max_version);
+        all_match = all_match && (all_packs[i].size() == dmfiles[i]->getPacks());
+        some_packs_count += some_packs[i]->size();
+    }
+
+    if (all_match)
+    {
+        LOG_DEBUG(log, "{} all match, total_rows={}, cost={}ms", __func__, segment_snap->stable->getDMFilesRows(), sw.elapsedMilliseconds());
+        return std::make_shared<BitmapFilter>(segment_snap->stable->getDMFilesRows(), segment_snap, /*default_value*/ true);
+    }
+
+    auto bitmap_filter = std::make_shared<BitmapFilter>(segment_snap->stable->getDMFilesRows(), segment_snap, /*default_value*/ false);
+    UInt32 preceded_dmfile_rows = 0;
+    for (size_t i = 0; i < dmfiles.size(); i++)
+    {
+        const auto & all_pack = all_packs[i];
+        for (const auto & pack : all_pack)
+        {
+            bitmap_filter->set(pack.offset + preceded_dmfile_rows, pack.rows);
+        }
+        preceded_dmfile_rows += dmfiles[i]->getRows();
+    }
+    if (some_packs_count <= 0)
+    {
+        return bitmap_filter;
+    }
+    static ColumnDefines columns_to_read{
+        getExtraHandleColumnDefine(is_common_handle),
+        getVersionColumnDefine(),
+        getTagColumnDefine(),
+    };
+    BlockInputStreamPtr stream = segment_snap->stable->getInputStream(dm_context,
+                                                                      columns_to_read,
+                                                                      read_ranges,
+                                                                      filter,
+                                                                      max_version,
+                                                                      expected_block_size,
+                                                                      /*enable_handle_clean_read*/ false,
+                                                                      /*is_fast_scan*/ false,
+                                                                      /*enable_del_clean_read*/ false,
+                                                                      /*read_packs*/ some_packs,
+                                                                      /*need_row_id*/ true);
```

Comment:
We need to fill `segment_row_id_col` with continuous row_id, and the filter twice. We can call `Block read(FilterPtr, true)` both in `DMRowKeyFilterBlockInputStream` and `DMVersionFilterBlockInputStream`, which means that there is no need to fill `segment_row_id_col` with continuous row_id, and filter only once.

Reply:
Since there is no `DeltaMergeBlockInputStream`, the order of rows will not change, so we can do this optimization.

Reply:
Are there some special reason for not supporting `read(FilterPtr, true)` in `DMRowKeyFilterBlockInputStream`? Maybe   we can achieve it since filter is heavy-cost?

Reply:
I think it is worth to do this now. We do not need to `createSegmentRowIdCol` in `ConcatSkippableBlockInputStream`. When the parent of `ConcatSkippableBlockInputStream` is `DeltaMergeBlockInputStream`, there is no need to `createSegmentRowIdCol` since  `DeltaMergeBlockInputStream` will do that. And the parent of `ConcatSkippableBlockInputStream` is `DMRowKeyFilterBlockInputStream`, we can call `read(FilterPtr, true)` to return filter, and we can use filter to generaate bitmap.

Reply:
no, I mean use filter to replace `SegmentRowIdCol`.

`ConcatSkippableBlockInputStream` return the following block:

```
startOffset: 100
col filter
1   1
1   1
2   1
3   1
4   1
5   1
6   1
```

After `DMVersionFilterBlockInputStream`, the block is:

```
startOffset: 100
col filter
1   0
1   1
2   1
3   1
4   1
5   1
6   1
```

After `DMRowKeyFilterBlockInputStream`, the block is:

```
startOffset: 100
col filter
1   0
1   0
2   1
3   1
4   1
5   1
6   0
```

Then we can just set [102, 103, 104, 105] of bitmap to 1.

Reply:
> However, read continuous packs is not necessary since the packs to be read has been specified. We can read multiple packs from DMFile concurrently in this situation in the future or in the cloud native mode, we can return several already cached packs first.

But fill `SegmentRowIdCol` in `ConcatSkippableBlockInputStream` can not handle this case either. Like we can a block in `ConcatSkippableBlockInputStream`, which contain data in `pack2`, `pack9` and `pack88`, how we can know about that?

Reply:
> because it would introduce special code paths and more restrictions but the benefits are small.

fill `SegmentRowIdCol` in `ConcatSkippableBlockInputStream` also introduce code path 🤔 And using `offset` only need little changes in `BitmapFilter`